kernel/thread.c

```diff
@@ -433,6 +433,12 @@ k_tid_t z_impl_k_thread_create(struct k_thread *new_thread,
 {
 	__ASSERT(!z_is_in_isr(), "Threads may not be created in ISRs");
 
+	/* Special case, only for unit tests */
+#if defined(CONFIG_TEST) && defined(CONFIG_ARCH_HAS_USERSPACE) && !defined(CONFIG_USERSPACE)
+	__ASSERT((options & K_USER) == 0,
+		 "Platform is capable of user mode, and test thread created with K_USER option, but CONFIG_TEST_USERSPACE or CONFIG_USERSPACE is not set\n");
```

Comment:
Wouldn't it be cleaner to just make it a general assertion if the caller asks for K_USER but the kernel can't accommodate?  Or do we document that as a valid noop?  Seems like there's nothing really test-specific about this validation other than the reason it was added.

Reply:
For general usage, it's valid to disable userspace and still use things like K_USER or other related APIs. They are no-ops.

However, just for tests, we want to tighten this up as if someone is dropping a test to user mode in a test case, but forgets to set CONFIG_TEST_USERSPACE, it will not silently just run everything in user mode.

We could consider making this a general assertion, but I am hesitant to do this here.